### PR TITLE
Ensure defined attributes do not overwrite existing methods

### PR DESCRIPTION
**Pull request by msalzburg**
Because overwriting existing methods with component attributes or elements is bad.

### CODE REVIEW

File: lib/components/element.rb [hunk 1]
```diff
@@ -5,6 +5,8 @@ def self.attributes
     end
 
     def self.attribute(name, default: nil)
+      raise_if_method_exists(name)
```
Comment (jensljungblad) [inline lib/components/element.rb]: Would it be weird to do a `define_method_or_raise_if_exists` or something? So it's baked into the define method call. Almost feels like there should be an option for `define_method` to do this.. :)

```ruby
def define_method_or_raise(method)
  raise(Components::Error, "Method '#{method}' already exists.") if respond_to?(method.to_sym)

  define_method(method) do
    yield
  end
end
```

File: lib/components/element.rb [hunk 2]
```diff
@@ -50,6 +56,14 @@ def self.element(name, multiple: false, &config)
     # rubocop:enable Metrics/MethodLength
     # rubocop:enable Metrics/PerceivedComplexity
 
+    class << self
```
Comment (jensljungblad) [inline lib/components/element.rb]: Should we move the other class methods here as well? I don't particularly care either way.

Comment (msalzburg) [reply lib/components/element.rb]: I think nesting all class methods within a `class << self; end` definition makes the code less readable. To define the private class method similar to the other class methods I now used the `private_class_method :xxx` definition.

File: lib/components/element.rb (hunk 1 above)
Comment (msalzburg) [reply lib/components/element.rb]: I like you thinking - refactored accordingly. There might be an issue if somebody catches our exception and is left with an "unstable" component instance - where the new attribute/element configuration is already stored in the components class instance variable but no method is defined for the attribute/element though.

Comment (jensljungblad) [reply lib/components/element.rb]: Hm, good point. Not sure how important that is..

If you want to put the raise back before method definitions perhaps we could do:

```ruby
def self.element(name, multiple: false, &config)
  raise_if_method_exists(name)
  plural_name = name.to_s.pluralize.to_sym if multiple
  raise_if_method_exists(plural_name)
  # ...
end
```

So we raise right away, before we've defined the singular method.

Comment (msalzburg) [reply lib/components/element.rb]: I guess we are fine for now. It is quite speculative that somebody is rescuing component exceptions ;) Lets merge and go on ...

**Comment by jensljungblad** [comment]
Looks great! Thanks for the contribution :)

**Pull request merged**